userland/Makefile

```diff
@@ -1,356 +1,64 @@
 # More info on our approach here: http://stackoverflow.com/questions/97338
-$(BUILDDIR)/%.o: %.c | $(BUILDDIR)
-	$(TRACE_DEP)
-	$(Q)$(CC) $(CFLAGS) $(CPPFLAGS) -MF"$(@:.o=.d)" -MG -MM -MP -MT"$(@:.o=.d)@" -MT"$@" "$<"
-	$(TRACE_CC)
-	$(Q)$(CC) $(CFLAGS) $(CPPFLAGS) -c -o $@ $<
-
-$(BUILDDIR)/%.o: %.cc | $(BUILDDIR)
-	$(TRACE_DEP)
-	$(Q)$(CXX) $(CXXFLAGS) $(CPPFLAGS) -MF"$(@:.o=.d)" -MG -MM -MP -MT"$(@:.o=.d)@" -MT"$@" "$<"
-	$(TRACE_CXX)
-	$(Q)$(CXX) $(CXXFLAGS) $(CPPFLAGS) -c -o $@ $<
-
-$(BUILDDIR)/%.o: %.cpp | $(BUILDDIR)
-	$(TRACE_DEP)
-	$(Q)$(CXX) $(CXXFLAGS) $(CPPFLAGS) -MF"$(@:.o=.d)" -MG -MM -MP -MT"$(@:.o=.d)@" -MT"$@" "$<"
-	$(TRACE_CXX)
-	$(Q)$(CXX) $(CXXFLAGS) $(CPPFLAGS) -c -o $@ $<
-
-
-# As different boards have different RAM/ROM sizes, we dynamically generate a
-# linker script (unless the user provide their own)
-LAYOUT ?= $(BOARD_BUILDDIR)/layout.ld
-
-# XXX(Pat) out of tree path to TOCK_BOARD directory? (?= chip hack)
-USERLAND_LAYOUT := $(TOCK_USERLAND_BASE_DIR)/userland_layout.ld
-CHIP_LAYOUT ?= $(TOCK_BASE_DIR)/boards/$(TOCK_BOARD)/chip_layout.ld
-
-$(BOARD_BUILDDIR)/layout.ld:	$(USERLAND_LAYOUT) $(CHIP_LAYOUT) | $(BOARD_BUILDDIR)
-	$(Q)echo "INCLUDE $(CHIP_LAYOUT)" > $@
-	$(Q)echo "INCLUDE $(USERLAND_LAYOUT)" >> $@
-
-
-.PHONY:	all
-all:	$(BOARD_BUILDDIR)/app.bin size
-
-.PHONY: size
-size:	$(BOARD_BUILDDIR)/app.elf
-	@$(SIZE) $<
-
-.PHONY: debug
-debug:	$(BOARD_BUILDDIR)/app.lst
-
-$(BOARD_BUILDDIR)/app.lst: $(BOARD_BUILDDIR)/app.elf
-	$(TRACE_LST)
-	$(Q)$(OBJDUMP) $(OBJDUMP_FLAGS) $< > $(BOARD_BUILDDIR)/app.lst
-
-# Include the libtock makefile. Adds rules that will rebuild library when needed
-include $(TOCK_USERLAND_BASE_DIR)/libtock/Makefile
+$(BUILDDIR)/$(1)/%.o: %.c | $(BUILDDIR)/$(1)
+	$$(TRACE_DEP)
+	$(Q)$(CC) $(CFLAGS) -mcpu=$(1) $(CPPFLAGS) -MF"$$(@:.o=.d)" -MG -MM -MP -MT"$$(@:.o=.d)@" -MT"$$@" "$$<"
+	$$(TRACE_CC)
+	$(Q)$(CC) $(CFLAGS) -mcpu=$(1) $(CPPFLAGS) -c -o $$@ $$<
+
+$(BUILDDIR)/$(1)/%.o: %.cc | $(BUILDDIR)/$(1)
+	$$(TRACE_DEP)
+	$(Q)$(CXX) $(CXXFLAGS) -mpcu=$(1) $(CPPFLAGS) -MF"$$(@:.o=.d)" -MG -MM -MP -MT"$$(@:.o=.d)@" -MT"$$@" "$$<"
+	$$(TRACE_CXX)
+	$(Q)$(CXX) $(CXXFLAGS) -mpcu=$(1) $(CPPFLAGS) -c -o $$@ $$<
+
+$(BUILDDIR)/$(1)/%.o: %.cpp | $(BUILDDIR)/$(1)
+	$$(TRACE_DEP)
+	$(Q)$(CXX) $(CXXFLAGS) -mcpu=$(1) $(CPPFLAGS) -MF"$$(@:.o=.d)" -MG -MM -MP -MT"$$(@:.o=.d)@" -MT"$$@" "$$<"
+	$$(TRACE_CXX)
+	$(Q)$(CXX) $(CXXFLAGS) -mpcu=$(1) $(CPPFLAGS) -c -o $$@ $$<
 
-$(BOARD_BUILDDIR)/app.elf: $(OBJS) $(TOCK_USERLAND_BASE_DIR)/newlib/libc.a $(LIBTOCK) $(LAYOUT) | $(BOARD_BUILDDIR)
-	$(TRACE_LD)
-	$(Q)$(CC) $(CFLAGS) $(CPPFLAGS)\
+# Collect all desired built output.
+$(BUILDDIR)/$(1)/$(1).elf: $(patsubst %.c,$(BUILDDIR)/$(1)/%.o,$(C_SRCS)) $(patsubst %.cc,$(BUILDDIR)/$(1)/%.o,$(filter %.cc, $(CXX_SRCS))) $(patsubst %.cpp,$(BUILDDIR)/$(1)/%.o,$(filter %.cpp, $(CXX_SRCS))) $(TOCK_USERLAND_BASE_DIR)/newlib/libc.a $(LIBTOCK) $(LAYOUT) | $(BUILDDIR)
```

Comment:
This is that `OBJS` problem I told you about. I would very much prefer it as intermediate variables, but don't know how to make that work with this new auto-generated rules method